### PR TITLE
add rolling options for rolling median and quantile

### PR DESCRIPTION
**Pull request by marcvanheerden**
Work in progress, still need to: 

- add more tests
- add rolling options for rolling_apply
- figure out where to put enum QuantileInterpolOptions

@ritchie46 on the last point, could you please make a recommendation? 
I didn't want to put it in polars-core because it's needed in polars-arrow but if it's not in polars-core then py03 doesn't implement PyClass for this enum.

Thanks

### CODE REVIEW

**Comment by ritchie46** [comment]
> @ritchie46 on the last point, could you please make a recommendation?
I didn't want to put it in polars-core because it's needed in polars-arrow but if it's not in polars-core then py03 doesn't implement PyClass for this enum.

I don't think pyo3 implements pyclass for this anyway? Maybe we can export in in polars prelude using `pub use polars_arrow::QuantileInterpolOptions`.

File: polars/polars-arrow/src/kernels/rolling/no_nulls.rs [hunk 1]
```diff
@@ -73,6 +88,79 @@ where
...
+    let mut buf: Vec<(T, f64)> = Vec::with_capacity(window_size);
+    let len = values.len();
+    let out = (0..len)
+        .map(|idx| {
+            let (start, end) = det_offsets_fn(idx, window_size, len);
+            let vals = unsafe { values.get_unchecked(start..end) };
+            buf = vals
+                .iter()
+                .zip(weights.iter())
+                .map(|(&x, &y)| (x, y))
```
Comment (ritchie46) [inline polars/polars-arrow/src/kernels/rolling/no_nulls.rs:149]: Is there a reason we don't multiply the weights here directly? Then we can take the same `Fa` as in `rolling_apply_weighted`:  `Fn(&[T], f64, QuantileInterpolOptions) -> f64`

File: polars/polars-arrow/src/kernels/rolling/no_nulls.rs [hunk 2]
```diff
@@ -95,13 +183,284 @@ where
...
+    let mut vals: Vec<f64> = values
+        .iter()
+        .copied()
+        .map(|x| NumCast::from(x).unwrap())
+        .collect();
+    vals.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+    let length = vals.len();
+
+    let mut idx = match interpolation {
```
Comment (ritchie46) [inline polars/polars-arrow/src/kernels/rolling/no_nulls.rs:332]: Maybe we can do the `idx` and the `f64` computation in a single pattern match? I think that will be a bit more performant.

File: polars/polars-arrow/src/kernels/rolling/nulls.rs [hunk 3]
```diff
@@ -58,6 +60,67 @@ where
...
+                Some(val) => val,
+                None => {
+                    validity.set(idx, false);
+                    f64::default()
+                }
+            }
+        })
+        .collect_trusted::<Vec<f64>>();
+
+    Arc::new(PrimitiveArray::from_data(
```
Comment (ritchie46) [inline polars/polars-arrow/src/kernels/rolling/nulls.rs:117]: I think we should make this generic over the `Float` trait. We expect rolling functions over `f32` to stay of that dtype.

**Comment by ritchie46** [comment]
@marcvanheerden Thanks a lot for this. It looks good already and this was not the easiest to pick up. I've left some comments.

Not all have to be addressed in this PR we could also iterate on this.

**Comment by marcvanheerden** [comment]
Thanks for the feedback, will finish this off tonight

File: polars/polars-arrow/src/kernels/rolling/no_nulls.rs (hunk 1 above)
Comment (marcvanheerden) [reply polars/polars-arrow/src/kernels/rolling/no_nulls.rs:149]: I interpreted weighted quantiles to be like [this definition](https://en.wikipedia.org/wiki/Weighted_median) of the weighted median. In this definition the order is preserved but the quantiles are just shifted around by the weights. Multiplying in the weights would potentially change the order. For example for values (1, 2, 3) and weights (0.97, 0.02, 0.01), if we multiply the weights in then the order reverses.

Happy to do it that way if you want but the use case for that is not clear to me

Comment (ritchie46) [reply polars/polars-arrow/src/kernels/rolling/no_nulls.rs:149]: The weights are intended to scale the window. For instance you could create a triangular window like so `1, 0.5, 0`. 

I think we should keep the weights interpretation the same for all rolling methods for consistency. 

We can add a weighted median/quantile later that takes separate weights for that operation.

Comment (marcvanheerden) [reply polars/polars-arrow/src/kernels/rolling/no_nulls.rs:149]: Okay great, will do

**Comment by marcvanheerden** [comment]
having a power failure tonight so will be a bit delayed with this

File: polars/polars-arrow/src/kernels/rolling/nulls.rs (hunk 3 above)
Comment (marcvanheerden) [reply polars/polars-arrow/src/kernels/rolling/nulls.rs:118]: This should be resolved, I didn't use the Float trait per se but the outcome is the same

File: polars/polars-arrow/src/kernels/rolling/no_nulls.rs (hunk 1 above)
Comment (marcvanheerden) [reply polars/polars-arrow/src/kernels/rolling/no_nulls.rs]: this is done now

**Comment by marcvanheerden** [comment]
Still working on:
- writing tests
- fixing the py03 issue
- the one unresolved comment

**Comment by marcvanheerden** [comment]
still struggling a bit with the pyo3 issue, will ask for help on the discord

File: polars/polars-arrow/src/kernels/rolling/no_nulls.rs (hunk 2 above)
Comment (marcvanheerden) [reply polars/polars-arrow/src/kernels/rolling/no_nulls.rs:353]: I'll make this change in my next PR if that's okay

Comment (ritchie46) [reply polars/polars-arrow/src/kernels/rolling/no_nulls.rs:353]: Sure!

**Comment by marcvanheerden** [comment]
@ritchie46 thanks for your help, I'll make a separate PR for rolling_apply since that is quite different. Let me know if you want any other changes here

File: polars/polars-arrow/src/kernels/rolling/no_nulls.rs [hunk 4]
```diff
@@ -73,6 +89,75 @@ where
...
+            buf.iter_mut()
+                .zip(vals.iter().zip(weights))
+                .for_each(|(b, (v, w))| *b = *v * *w);
+
+            aggregator(&buf, quantile, interpolation)
+        })
+        .collect_trusted::<Vec<f64>>();
+
+    let validity = create_validity(min_periods, len as usize, window_size, det_offsets_fn);
+    Arc::new(PrimitiveArray::from_data(
```
Comment (ritchie46) [inline polars/polars-arrow/src/kernels/rolling/no_nulls.rs:154]: This also should work on both `f32` and `f64` data. I think you can use the same solution as you did on polars/polars-arrow/src/kernels/rolling/nulls.rs  line 63

**Comment by ritchie46** [comment]
Thanks a lot @marcvanheerden its almost good to go. I only have one remaining comment. A kernel that only returns a `Float64` that should also return `Float32` it its of that dtype.

**Comment by marcvanheerden** [comment]
Thanks, just want to clarify, I think this functionality is currently working but it's in happening in polars-core/src/chunked_array/ops/rolling_window.rs . Are you saying it to rather the type changes to be handled in polars-arrow/src/kernels/rolling/ ?

**Comment by ritchie46** [comment]
> Thanks, just want to clarify, I think this functionality is currently working but it's in happening in polars-core/src/chunked_array/ops/rolling_window.rs . Are you saying the type changes should rather be handled in polars-arrow/src/kernels/rolling/ ?

I am on mobile atm. But we basically need an implementation for f32 and f64 and f32 arrays dispatch to the f32 implementation and the rest to the f64 (somerimes with cast)

**Comment by marcvanheerden** [comment]
Okay, I tested it and this is how it is currently working. the rolling_median and rolling_quantile function in polars-core cast to f64 and recurse if the input is not a float. That way f32 -> f32 and anything else comes out as f64. I'll add a test to confirm

**Comment by ritchie46** [comment]
> Okay, I tested it and this is how it is currently working. the rolling_median and rolling_quantile function in polars-core cast to f64 and recurse if the input is not a float. That way f32 -> f32 and anything else comes out as f64. I'll add a test to confirm

Given that `rolling_apply_convolve_quantile` is not generic over floats it needs f32 to be cast to f64 to use that kernel. That cast can double memory and is expensive.

**Comment by marcvanheerden** [comment]
Ah. apologies, I missed that. Will fix and write some more tests today

**Comment by marcvanheerden** [comment]
I think it's good to go now, thanks for your help

**Comment by ritchie46** [comment]
> I think it's good to go now, thanks for your help

And thank you for all the effort. It's a good addition!

**Pull request merged**